File: output/java/Easy/105-find-mode-in-binary-search-tree.java
/**https://leetcode.com/problems/find-mode-in-binary-search-tree/ */
//Given the root of a binary search tree (BST) with duplicates, return all the mode(s) (i.e., the most frequently occurred element) in it.
//If the tree has more than one mode, return them in any order.
//Assume a BST is defined as follows:
//The left subtree of a node contains only nodes with keys less than or equal to the node's key.
//The right subtree of a node contains only nodes with keys greater than or equal to the node's key.
//Both the left and right subtrees must also be binary search trees.


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class FindModeInBinarySearchTree {
    public int[] findMode(TreeNode root) {
        if(root == null) return new int[0];
        Map<Integer, Integer> map = new HashMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()){
            TreeNode node = queue.poll();
            map.put(node.val, map.getOrDefault(node.val, 0) + 1);
            if(node.left != null) queue.offer(node.left);
            if(node.right != null) queue.offer(node.right);
        }
        int max = 0;
        for(int key : map.keySet()){
            max = Math.max(max, map.get(key));
        }
        List<Integer> list = new ArrayList<>();
        for(int key : map.keySet()){
            if(map.get(key) == max){
                list.add(key);
            }
        }
        int[] ans = new int[list.size()];
        for(int i = 0; i < list.size(); i++){
            ans[i] = list.get(i);
        }
        return ans;
        
    }
}

    

File: output/c/Easy/105-find-mode-in-binary-search-tree.c
/**https://leetcode.com/problems/find-mode-in-binary-search-tree/ */
//Given the root of a binary search tree (BST) with duplicates, return all the mode(s) (i.e., the most frequently occurred element) in it.
//If the tree has more than one mode, return them in any order.
//Assume a BST is defined as follows:
//The left subtree of a node contains only nodes with keys less than or equal to the node's key.
//The right subtree of a node contains only nodes with keys greater than or equal to the node's key.
//Both the left and right subtrees must also be binary search trees.


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* findMode(struct TreeNode* root, int* returnSize){
    int* ans = (int*)malloc(10000 * sizeof(int));
    *returnSize = 0;
    int max = 0;
    int count = 0;
    int prev = 0;
    struct TreeNode* curr = root;
    struct TreeNode* pre = NULL;
    while(curr != NULL){
        if(curr->left == NULL){
            if(prev == curr->val){
                count++;
            }else{
                count = 1;
            }
            if(count > max){
                max = count;
                *returnSize = 0;
                ans[*returnSize] = curr->val;
                (*returnSize)++;
            }else if(count == max){
                ans[*returnSize] = curr->val;
                (*returnSize)++;
            }
            prev = curr->val;
            curr = curr->right;
        }else{
            pre = curr->left;
            while(pre->right != NULL && pre->right != curr){
                pre = pre->right;
            }
            if(pre->right == NULL){
                pre->right = curr;
                curr = curr->left;
            }else{
                pre->right = NULL;
                if(prev == curr->val){
                    count++;
                }else{
                    count = 1;
                }
                if(count > max){
                    max = count;
                    *returnSize = 0;
                    ans[*returnSize] = curr->val;
                    (*returnSize)++;
                }else if(count == max){
                    ans[*returnSize] = curr->val;
                    (*returnSize)++;
                }
                prev = curr->val;
                curr = curr->right;
            }
        }
    }
    return ans;

}

    